[libpas] Introduce stat-counter system for libpas

Achierius opened the pull request:
#### 1e34af6e2696cfee789e5dfc6322e10a30718b53
<pre>
[libpas] Introduce stat-counter system for libpas
<a href="https://bugs.webkit.org/show_bug.cgi?id=299311">https://bugs.webkit.org/show_bug.cgi?id=299311</a>
<a href="https://rdar.apple.com/160953463">rdar://160953463</a>

Reviewed by Dan Hecht.

This patch adds a new stat-counter system to libpas, with the intention
of tracking allocator-internal statistics and performance data. A
motivating example is to measure how many allocations are made per
size-class for the purposes of better tuning libpas&apos; configuration
parameters-- we&apos;ve done this in the past, but it&apos;s always been ad-hoc
and thus inefficient / less trustworthy than a built-in and maintained
system would be.

By default, these are compiled out of the build, and even when built in
they include a runtime configuration option (PAS_STATS_ENABLE) which can
be set to turn some/all stat-counters on,
e.g. `PAS_STATS_ENABLE=counter_a,counter_b`, or `PAS_STATS_ENABLE=1` to
turn on all available stat counters.

By default, stats are logged to stdout (as json blobs), but setting
`PAS_STATS_LOG_FILE` will instead log them out to the specified file.
Since not all environments in which libpas runs have a clean &apos;exit&apos; hook
(e.g. Safari tends to call terminate() in order to ensure it exits
quickly) this system instead logs all counters periodically  based on
the total number of stat-count-events which have taken place across all
threads. I tried doing this in the scavenger but that doesn&apos;t work very
well since the scavenger runs infrequently enough that we end up
missing a lot of counter events near the end of a test.

A rough json schema with the current counters:
```
{
  &quot;pid&quot;: &lt;INT&gt;,
  &quot;time_ns&quot;: &lt;INT&gt;,
  &quot;per_stat_data&quot;: {
    &quot;malloc_info_allocations&quot;: {
      &quot;total_count&quot;: &lt;INT&gt;,
      &quot;count_by_heap_type&quot;: [&lt;INT&gt;, &lt;INT&gt;, &lt;INT&gt;],
      &quot;count_by_size&quot;: [&lt;INT&gt;, &lt;INT&gt;, &lt;INT&gt;, &lt;INT&gt;, ..., &lt;INT&gt;],
    },
    &quot;malloc_info_bytes&quot;: {
      &lt;SAME AS FOR malloc_info_allocations&gt;
    }
  }
}
```

This is meant to be extensible and flexible enough to handle different
kinds of stat counters; I would have preferred to do this with
C++, and did look into doing so, but the interface layer ends up being
slower and un-ergonomic so I ultimately went with C to better fit in
with the rest of libpas. This means that we use everyone&apos;s favorite
macro-for-each (PAS_STATS_FOR_EACH_COUNTER) to define new counters.
I wasn&apos;t able to find a clean way to incorporate the definition of the
PAS_RECORD_STAT_&lt;statname&gt; shims, so registering a new counter does
require changes in two places -- but they&apos;re close by so I think it&apos;s
acceptable.

Since this framework is intended for use inside of an allocator, it is
intended to have low overhead (both for enabled and disabled counters) and to
have minimal use of heap-allocated memory -- however, there is room for
improvement on both counts.

Re.: heap-allocated memory:
on the logging path we currently do rely on heap allocations to make it easier
for people to add new counters, as using a fixed-size static allocation per
counter would mean every counter would need to pre-compute the theoretical
maximum size of its json payload.
Normally the utility heap would be a good fit for this use-case, but to avoid
reentrancy we do not use libpas to allocate this memory -- even by going
through the system heap. Instead, we call `malloc` directly. These buffers are
cached so it shouldn&apos;t happen often but it would be better to be able to remove
that dependency.
That exception notwithstanding, pas_stats.h does&apos;t malloc anything anywhere else,
so if we do figure out a way to improve this then we only need to change
`pas_stats_ensure_print_buffer` to match.

Re.: performance:
the current design is not bad but does introduce a lot of atomic traffic and
cross-core contention. Ideally, we would instead have a per-thread &apos;local stat
counter cache&apos; which we would then periodically accumulate into a global
stat-counter object. Individual stat counters would have to be aware since they
need to implement their own accumulate functions.
Even better than thread-local would be if we had something like Linux&apos; rseqs,
as we could then store this data per-CPU and avoid any migration whatsoever.
In both cases though, we would risk under-counting statistics unless we
implemented an analog of what pas-TLCs do where they iterate over other
threads&apos; TLCs and collect data out. Doing so generically across all kinds of
stat counters seems like a challenge.

Canonical link: <a href="https://commits.webkit.org/304490@main">https://commits.webkit.org/304490@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c13979fd666217c431f419311c1aa919fe2c7263

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135744 "13 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8120 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47037 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143458 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/612d1b11-98cd-4ff0-bdb7-acd1a9fcd6f2) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8777 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7967 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103737 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c43389f6-ce4e-458d-b7b8-dfe90b1c4013) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138690 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/6325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/121675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84613 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c9975815-815a-49e0-a718-0b72a5c2492f) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4064 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127726 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/39862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146207 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134227 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7809 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/40430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7835 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/6553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112480 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28533 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/5952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/117975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7855 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/36083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167048 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/7592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7816 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/7689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,307 @@
+// buf and len are always owned by pas_stats_ensure_print_buffer, and the
+// underlying storage should be considered an implementation detail.
+typedef struct pas_stats_print_buffer {
+    char* buf;
```
nit: it's little unfortunate to call this field `buf` when the field in pas_stats_state_base is also called `buf`. Maybe call this one `ptr`, so then it's `buf.ptr` & `buf.len`?

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,307 @@
+
+extern pas_stats_data g_pas_stats_data;
+
+#define PAS_STATS_LOG_INTERVAL (1 << 16)
```
To make the units of this more obvious, maybe call it PAS_STATS_LOG_COUNT_INTERVAL or something?

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_bitfit_page_inlines.h:
```diff
@@ -247,6 +248,7 @@ static PAS_ALWAYS_INLINE pas_bitfit_allocation_result pas_bitfit_page_finish_all
     pas_bitfit_page_testing_verify(page);
 
     PAS_PROFILE(BITFIT_ALLOCATION, &page_config, begin, size, allocation_mode);
+    PAS_RECORD_STAT_MALLOC(pas_stats_heap_type_bitfit, size);
```
Why not just use the PAS_PROFILE hook rather than adding an new hook?

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+typedef struct pas_stats_sink pas_stats_sink;
+typedef void (*pas_stats_sink_output_func)(pas_stats_sink* sink, const char* json_output);
+
+// buf and len are always owned by pas_stats_ensure_print_buffer, and the
```
Need update?  "buf and len" => "ptr and len".

Also, WebKit style is to not use abbreviations like len.  There are a few places where this has escaped into the codebase, but let's try to stay with the expected style as best as we can.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+char* pas_stats_ensure_print_buffer(pas_stats_print_buffer*, size_t desired_len);
+
+// All stat-counter structs must have a field of this type named '.data' as
+// their first member
```
Is this reference to '.data' stale?  I don't see any reference to a `data` field anywhere in the patch.

Achierius replied:
Not stale, I just mistyped this for whatever reason

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+/* General stat-counter machinery */
+/**********************************/
+
+typedef char* (*pas_stats_stat_logger)(void* stat);
```
nit: let's call this `pas_stats_stat_logger_function`.  I think that helps makes it clear that this is a function pointer and not a struct when used as a fields as is the case in the definition of `pas_stats_stat_base` below.

You can also shorten it to `pas_stats_logger_function`.  I'm not sure the extra `stat` adds much context.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+typedef struct {
+    const char* name;
+    pas_stats_stat_logger logger;
+    pas_stats_print_buffer buf;
```
Let's use `buffer` instead of `buf`.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+void pas_stats_log_all_enabled_stats(void);
+
+// It's OK for the .enabled check here to be non-atomic: we know that all
+// .enabled bits will start initialized to 1, and will be set to 0 at most once.
```
Since `.enabled` is of type `bool`, let's be consistent and use `true` and `false` everywhere instead of `1` and `0`.  I see that it is initialized to `true`, but other places in the code sets it to `1` or `0`.  Let's fix all those places too.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+// some unecessary atomic writes at the beginning, but it avoids placing an
+// extra pthread_once check inline with every statistic, both enabled and
+// disabled.
+void pas_stats_setup_if_necessary(void);
```
Since `pas_stats_setup_if_necessary()` is only ever called from `pas_stats_do_accounting_before_recording_stat_slow_path()`, let's make it a static function in the .c file instead.  Is there any reason to declare it publicly like this?

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+
+#define PAS_RECORD_STAT_MALLOC(heap_type, size) \
+    PAS_RECORD_STAT(malloc_info_bytes, heap_type, size); \
+    PAS_RECORD_STAT(malloc_info_allocations, heap_type, size);
```
This is a bug.  `PAS_RECORD_STAT_MALLOC` should enclose its 2 `PAS_RECORD_STAT` invocations in a `do-while(false)` statement.  Otherwise, clients that do this:
```
    if (result.did_succeed)
        PAS_RECORD_STAT_MALLOC(pas_stats_heap_type_large, size);
```
... while have the condition apply to the 1st `PAS_RECORD_STAT` but not the 2nd.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+    pas_stats_malloc_info_record(data, heap_type, size, 1)
+
+#else /* !PAS_ENABLE_STATS */
+#define PAS_RECORD_STAT(name, ...) do { PAS_UNUSED_V(__VA_ARGS__); } while (0)
```
Please move this to the same `#if PAS_ENABLE_STATS` block below where `PAS_RECORD_STAT` is defined.  It is very disorienting and hard to read that the disabled version is here, while the enabled version is elsewhere.  Let's keep them paired together.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+    bool enabled;
+} pas_stats_stat_base;
+
+/******************************************************************/
+/* Implementations for the individual stat-counter instantiations */
+/******************************************************************/
```
nit: Maybe move this whole section to the bottom of the file so that we're not interrupting the "general stat-counter machinery" section here.

Achierius replied:
Unfortunately it's not possible since this section needs to know the definition of `pas_stats_stat_base`; it's also not possible to lift the other half of 'general stat-counter machinery' up since it needs to know the definitions of the per-stat counter structs.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,311 @@
+// counters will never actually be used since pas_stats_log_all_enabled_stats
+// does use an atomic check for whether .enabled is set.
+#define PAS_RECORD_STAT(name, ...) do { \
+        if (g_pas_stats_data.per_stat_data. name .base.enabled) { \
+            pas_stats_do_accounting_before_recording_stat(); \
```
Instead of incrementing `g_pas_stats_data.log_counter` once per stat type, let's add a bool in `pas_stats_data` which is set to true if any of the underlying stats is true.  Then, have `PAS_RECORD_STAT_MALLOC` check this bool and increment `g_pas_stats_data.log_counter` once if needed.  Do this before invoking the respective `PAS_RECORD_STAT`s.  Each `PAS_RECORD_STAT` need not call `pas_stats_do_accounting_before_recording_stat()` themselves.  This will minimize the perf cost when no stats collection is enabled.

Achierius replied:
I expect most other stat-counters to be 'solitary', i.e. incremented on their own -- at least among the ones I have in mind to add. If that changes I'll want to create a proper 'stat group' layer over the base data level, but until then I don't want to complicate the code too much in service of 'paired' stats like this.

For this case, I think we can achieve the same perf savings by adding a PAS_RECORD_STAT_WITHOUT_ACCOUNTING macro which would skip the accounting step, and then making PAS_RECORD_STAT_MALLOC use that for the second call.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.c:
```diff
@@ -0,0 +1,466 @@
+
+    uint64_t log_time = current_time_ns();
+    if (!g_pas_stats_data.first_log_time_ns)
+        g_pas_stats_data.first_log_time_ns = current_time_ns();
```
You're not using `first_log_time_ns` for anything other than to serve as an epoch.

Let's rename it to `start_time_ns` and initialize this once in `pas_stats_setup()` instead.  There's no value add in doing this check every time.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.c:
```diff
@@ -0,0 +1,466 @@
+    log_time -= g_pas_stats_data.first_log_time_ns;
+
+    if (!g_pas_stats_data.pid)
+        g_pas_stats_data.pid = getpid();
```
Ditto.  Do this init just once in `pas_stats_setup()` instead.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_stats.c:
```diff
@@ -0,0 +1,466 @@
+#define COUNT_ONE(name, type, logger_sym) + 1
+        PAS_STATS_FOR_EACH_COUNTER(COUNT_ONE)
+#undef COUNT_ONE
+    };
```
nit: for readability, I suggest re-expressing this as:
```
#define COUNT_ONE(name, type, logger_sym) + 1
    enum {
        pas_stats_count = 0 + PAS_STATS_FOR_EACH_COUNTER(COUNT_ONE)
    };
#undef COUNT_ONE
```
Actually, is there any reason you can't express this enum as a `const unsigned` instead?
```
    const unsigned pas_stats_count = 0 + PAS_STATS_FOR_EACH_COUNTER(COUNT_ONE)'
```

Achierius replied:
Did the latter

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_stats.c:
```diff
@@ -0,0 +1,513 @@
+/*
+ * Copyright (c) 2024 Apple Inc. All rights reserved.
```
2025

Achierius replied:
Fixed

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright (c) 2024 Apple Inc. All rights reserved.
```
2025

Achierius replied:
Fixed

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,381 @@
+            PAS_RECORD_STAT_ ## name(&(g_pas_stats_data.per_stat_data. name ), __VA_ARGS__); \
+        } \
+    } while (0)
+// Since this version does not call any setup/logging functions, it should only
```
What's the motivation for not doing the setup/logging in this version? I think future readers will wonder this too and will consider adding that unless you give a reason.

Achierius replied:
It's a to request from Mark to reduce performance overhead, from a review comment earlier in this PR
I'll mention as such

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_stats.h:
```diff
@@ -0,0 +1,381 @@
+// be called if you are A) ok with this stat not being logged unless other stats
+// trigger a logging pass on their own, B) sure that this stat will only be
+// incremented after at least one other stat has called
+// pas_stats_do_accounting_before_recording_stat().
```
Does this mean you should be okay with A or B, or you must be okay with A and B? If the latter, I think it's the former, in which case I'd put the word "or" before B) to be explicit.

Achierius replied:
Added "or"

dhecht reviewed (approved):
r=me with comments

Achierius committed:
[libpas] Introduce stat-counter system for libpas
https://bugs.webkit.org/show_bug.cgi?id=299311
rdar://160953463

Reviewed by Dan Hecht.

This patch adds a new stat-counter system to libpas, with the intention
of tracking allocator-internal statistics and performance data. A
motivating example is to measure how many allocations are made per
size-class for the purposes of better tuning libpas' configuration
parameters-- we've done this in the past, but it's always been ad-hoc
and thus inefficient / less trustworthy than a built-in and maintained
system would be.

By default, these are compiled out of the build, and even when built in
they include a runtime configuration option (PAS_STATS_ENABLE) which can
be set to turn some/all stat-counters on,
e.g. `PAS_STATS_ENABLE=counter_a,counter_b`, or `PAS_STATS_ENABLE=1` to
turn on all available stat counters.

By default, stats are logged to stdout (as json blobs), but setting
`PAS_STATS_LOG_FILE` will instead log them out to the specified file.
Since not all environments in which libpas runs have a clean 'exit' hook
(e.g. Safari tends to call terminate() in order to ensure it exits
quickly) this system instead logs all counters periodically  based on
the total number of stat-count-events which have taken place across all
threads. I tried doing this in the scavenger but that doesn't work very
well since the scavenger runs infrequently enough that we end up
missing a lot of counter events near the end of a test.

A rough json schema with the current counters:
```
{
  "pid": <INT>,
  "time_ns": <INT>,
  "per_stat_data": {
    "malloc_info_allocations": {
      "total_count": <INT>,
      "count_by_heap_type": [<INT>, <INT>, <INT>],
      "count_by_size": [<INT>, <INT>, <INT>, <INT>, ..., <INT>],
    },
    "malloc_info_bytes": {
      <SAME AS FOR malloc_info_allocations>
    }
  }
}
```

This is meant to be extensible and flexible enough to handle different
kinds of stat counters; I would have preferred to do this with
C++, and did look into doing so, but the interface layer ends up being
slower and un-ergonomic so I ultimately went with C to better fit in
with the rest of libpas. This means that we use everyone's favorite
macro-for-each (PAS_STATS_FOR_EACH_COUNTER) to define new counters.
I wasn't able to find a clean way to incorporate the definition of the
PAS_RECORD_STAT_<statname> shims, so registering a new counter does
require changes in two places -- but they're close by so I think it's
acceptable.

Since this framework is intended for use inside of an allocator, it is
intended to have low overhead (both for enabled and disabled counters) and to
have minimal use of heap-allocated memory -- however, there is room for
improvement on both counts.

Re.: heap-allocated memory:
on the logging path we currently do rely on heap allocations to make it easier
for people to add new counters, as using a fixed-size static allocation per
counter would mean every counter would need to pre-compute the theoretical
maximum size of its json payload.
Normally the utility heap would be a good fit for this use-case, but to avoid
reentrancy we do not use libpas to allocate this memory -- even by going
through the system heap. Instead, we call `malloc` directly. These buffers are
cached so it shouldn't happen often but it would be better to be able to remove
that dependency.
That exception notwithstanding, pas_stats.h does't malloc anything anywhere else,
so if we do figure out a way to improve this then we only need to change
`pas_stats_ensure_print_buffer` to match.

Re.: performance:
the current design is not bad but does introduce a lot of atomic traffic and
cross-core contention. Ideally, we would instead have a per-thread 'local stat
counter cache' which we would then periodically accumulate into a global
stat-counter object. Individual stat counters would have to be aware since they
need to implement their own accumulate functions.
Even better than thread-local would be if we had something like Linux' rseqs,
as we could then store this data per-CPU and avoid any migration whatsoever.
In both cases though, we would risk under-counting statistics unless we
implemented an analog of what pas-TLCs do where they iterate over other
threads' TLCs and collect data out. Doing so generically across all kinds of
stat counters seems like a challenge.

Canonical link: https://commits.webkit.org/304490@main
